src/core/lib/event_engine/grpc_ares_wrapper.cc

```diff
@@ -0,0 +1,1037 @@
+  // If dns_server is specified, use it.
+  absl::Status error = SetRequestDNSServer(dns_server);
+  if (!error.ok()) {
+    ares_destroy(channel_);
+    return error;
+  }
+  initialized_ = true;
+  return absl::OkStatus();
+}
+
+absl::Status GrpcAresRequest::SetRequestDNSServer(
+    absl::string_view dns_server) {
+  if (!dns_server.empty()) {
+    GRPC_ARES_WRAPPER_TRACE_LOG("request:%p Using DNS server %s", this,
+                                dns_server.data());
+    grpc_resolved_address addr;
+    struct ares_addr_port_node dns_server_addr;
+    if (grpc_parse_ipv4_hostport(dns_server, &addr, /*log_errors=*/false)) {
+      dns_server_addr.family = AF_INET;
+      struct sockaddr_in* in = reinterpret_cast<struct sockaddr_in*>(addr.addr);
+      memcpy(&dns_server_addr.addr.addr4, &in->sin_addr,
+             sizeof(struct in_addr));
+      dns_server_addr.tcp_port = grpc_sockaddr_get_port(&addr);
+      dns_server_addr.udp_port = grpc_sockaddr_get_port(&addr);
+    } else if (grpc_parse_ipv6_hostport(dns_server, &addr,
+                                        /*log_errors=*/false)) {
+      dns_server_addr.family = AF_INET6;
+      struct sockaddr_in6* in6 =
+          reinterpret_cast<struct sockaddr_in6*>(addr.addr);
+      memcpy(&dns_server_addr.addr.addr6, &in6->sin6_addr,
+             sizeof(struct in6_addr));
+      dns_server_addr.tcp_port = grpc_sockaddr_get_port(&addr);
+      dns_server_addr.udp_port = grpc_sockaddr_get_port(&addr);
+    } else {
+      return absl::Status(absl::StatusCode::kInvalidArgument,
+                          absl::StrCat("Cannot parse authority: ", dns_server));
+    }
+    // Prevent an uninitialized variable.
+    dns_server_addr.next = nullptr;
+    int status = ares_set_servers_ports(channel_, &dns_server_addr);
+    if (status != ARES_SUCCESS) {
+      return AresStatusToAbslStatus(
+          status, absl::StrCat("c-ares status is not ARES_SUCCESS: ",
+                               ares_strerror(status)));
+    }
+  }
+  return absl::OkStatus();
+}
+
+void GrpcAresRequest::OnReadable(FdNode* fd_node, absl::Status status) {
+  mu().Lock();
+  GPR_ASSERT(fd_node->readable_registered);
+  fd_node->readable_registered = false;
+  GRPC_ARES_WRAPPER_TRACE_LOG("OnReadable: fd: %d; request: %p; status: %s",
+                              fd_node->as, this, status.ToString().c_str());
+  PrintCurrentStackTrace();
+  if (status.ok() && !shutting_down_) {
+    do {
+      ares_process_fd(channel_, fd_node->as, ARES_SOCKET_BAD);
+    } while (fd_node->polled_fd->IsFdStillReadableLocked());
+  } else {
+    // If error is not absl::OkStatus() or the resolution was cancelled, it
+    // means the fd has been shutdown or timed out. The pending lookups made
+    // on this request will be cancelled by the following ares_cancel(). The
+    // remaining file descriptors in this request will be cleaned up in the
+    // following Work() method.
+    ares_cancel(channel_);
+  }
+  Work();
```

Comment:
Currently the contract of `Work` is that we need to check if `shutting_down_ == true` and then set up the user callback / completion if so.

What do you think about having `Work` itself do this?

I think that would make the lifetimes easier to reason about because it would be very clear that the user callback is invoked only after `set_shutdown` has been called AND after we've initiated shutdown on all possible file descriptors.

Reply:
In the new implementation, this part is simplified and the code after `Work` is no longer needed. The user callback is already triggered in the respective ares_callback.

Reply:
The code still seems to be here - are we missing commits?

Reply:
Oh sorry, I'm responding ahead of time since the next update will make these comments outdated (this file will be removed)...

Reply:
Actually, maybe, we should just get rid of that loop in the on_Readable callback entirely.

Then in CheckSocketsLocked, before registering an fd for an onreadable event, first check if there are still unread bytes remaining and just directly schedule the on_Readable callback if so